cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/AbstractConnectOperator.java

```diff
@@ -593,6 +603,184 @@ private Future<List<Condition>> maybeRestartConnectorTask(Reconciliation reconci
+                    LOGGER.warnCr(reconciliation, message);
+                    conditions.add(StatusUtils.buildWarningCondition("AlterOffsets", message));
+                    return Future.succeededFuture(conditions);
+                });
+    }
+
+    private Future<String> getOffsetsForAlterRequest(Map<String, Object> connectorStatus, String configMapNamespace, String configMapName, String configMapKeyName) {
+        @SuppressWarnings({ "rawtypes" })
+        Object path = ((Map) connectorStatus.getOrDefault("connector", emptyMap())).get("state");
+        if (!(path instanceof String state)) {
+            return Future.failedFuture("JSON response lacked $.connector.state");
+        }
+        if (!"STOPPED".equals(state)) {
+            return Future.failedFuture("Failed to alter the connector offsets, connector is not in STOPPED state");
+        }
+        return configMapOperations.getAsync(configMapNamespace, configMapName)
+                .compose(configMap -> {
+                    if (configMap == null) {
+                        return Future.failedFuture(String.format("Encountered error fetching offsets to use in alter operation. ConfigMap %s/%s does not exist.", configMapNamespace, configMapName));
+                    }
+                    if (configMap.getData().get(configMapKeyName) == null) {
+                        return Future.failedFuture(String.format("Encountered error fetching offsets to use in alter operation. Data field %s is missing.", configMapKeyName));
+                    }
+                    try {
+                        String offsets = configMap.getData().get(configMapKeyName);
+                        new ObjectMapper().readValue(offsets, Object.class);
+                        return Future.succeededFuture(offsets);
+                    } catch (IOException e) {
+                        return Future.failedFuture(String.format("Failed to parse contents of %s as JSON: %s", configMapKeyName, e.getMessage()));
+                    }
+                });
+    }
+
+    /**
+     * Resets the connector offsets.
+     * This operation requires the operator to be in a STOPPED state.
+     */
+    @SuppressWarnings({ "rawtypes" })
+    Future<List<Condition>> resetConnectorOffsets(Reconciliation reconciliation, String host, KafkaConnectApi apiClient, String connectorName, CustomResource resource, List<Condition> conditions) {
+        return apiClient.status(reconciliation, host, port, connectorName)
+                .compose(status -> {
+                    @SuppressWarnings({ "rawtypes" })
+                    Object path = ((Map) status.getOrDefault("connector", emptyMap())).get("state");
+                    if (!(path instanceof String state)) {
+                        return Future.failedFuture("JSON response lacked $.connector.state");
+                    }
+                    if (!"STOPPED".equals(state)) {
```

Comment:
Do we have some `Enum` instead of this magic constant? I assume you can use `io.strimzi.api.kafka.model.common.ConnectorState`? :) The same applies in previous parts.

Reply:
> The tricky part here is the ConnectorState enum values are based on the CRD field which is lowercase, whereas the state returned from the Connect API is upper case. @scholzj @mimaison what do you think of adding a fromStatus method to io.strimzi.api.kafka.model.common.ConnectorState that we can call directly here to convert the state to the ConnectorState enum?

I'm not sure I would add special method like that to the `api` module. But maybe you can simply convert it to lowercase in the `forValue` method? We do that in other places ... `ProcessRoles`, `NodeAddressType` etc.

> I noticed that the code styles dislikes using toUpperCase and toLowerCase

You should be able to use them, but you should specify the locale:

```java
toLowerCase(Locale.ENGLISH)
```

Reply:
Ah I misunderstood the code style error then, I've updated it to do the `toLowerCase`